lib/vector-core/src/event/log_event.rs

```diff
@@ -64,18 +138,23 @@ impl LogEvent {
     #[must_use]
     pub fn new_with_metadata(metadata: EventMetadata) -> Self {
         Self {
-            fields: Arc::new(Value::Object(Default::default())),
+            inner: Default::default(),
             metadata,
         }
     }
 
     ///  Create a `LogEvent` into a tuple of its components
     pub fn from_parts(map: BTreeMap<String, Value>, metadata: EventMetadata) -> Self {
-        let fields = Value::Object(map);
-        Self {
-            fields: Arc::new(fields),
-            metadata,
-        }
+        let inner = Arc::new(Inner::from(map));
+        Self { inner, metadata }
+    }
+
+    fn fields_mut(&mut self) -> &mut Value {
+        let result = Arc::make_mut(&mut self.inner);
+        // We MUST invalidate the inner size cache when making a
+        // mutable copy, since the _next_ action will modify the data.
+        result.invalidate();
+        &mut result.fields
```

Comment:
I really wish there was a way to determine if it actually did the clone, but I guess at worst it's an extra atomic load or two when the log event doesn't _have_ to clone.